sdks/python/apache_beam/io/gcp/bigtableio.py

```diff
@@ -141,3 +147,273 @@ def expand(self, pvalue):
+    :param start_key(byte): The start key row in the range.
+    :param end_key(byte): The end key row in the range.
+    '''
+    range_tracker = LexicographicKeyRangeTracker(start_key, end_key)
+    return self.split_range_subranges(current_size,
+                                      desired_bundle_size,
+                                      range_tracker)
+
+  def fraction_to_position(self, position, range_start, range_stop):
+    ''' We use the ``fraction_to_position`` method in
+    ``LexicographicKeyRangeTracker`` class to split a
+    range into two chunks.
+
+    :param position:
+    :param range_start:
+    :param range_stop:
+    :return:
+    '''
+    return LexicographicKeyRangeTracker.fraction_to_position(position,
+                                                             range_start,
+                                                             range_stop)
+
+  def split_range_subranges(self,
+                            sample_size_bytes,
+                            desired_bundle_size,
+                            ranges):
+    ''' This method split the range you get using the
+    ``desired_bundle_size`` as a limit size, It compares the
+    size of the range and the ``desired_bundle size`` if it is necessary
+    to split a range, it uses the ``fraction_to_position`` method.
+
+    :param sample_size_bytes: The size of the Range.
+    :param desired_bundle_size: The desired size to split the Range.
+    :param ranges: the Range to split.
+    '''
+    start_key = ranges.start_position()
+    end_key = ranges.stop_position()
+
+    split_ = float(desired_bundle_size)/float(sample_size_bytes)
+    split_count = int(math.ceil(split_))
+    portions = (1/split_)
+    if portions > 1:
+      for i in range(split_count):
+        estimate_position = ((i + 1) * split_)
+
+        for i in range(int(portions)):
+          temp_estimate = estimate_position*(i+1)
+          position = self.fraction_to_position(temp_estimate,
+                                               ranges.start_position(),
+                                               ranges.stop_position())
+          end_key = position
+          yield iobase.SourceBundle(sample_size_bytes * split_,
+                                    self,
+                                    start_key,
+                                    end_key)
+          start_key = position
+    else:
+      yield iobase.SourceBundle(sample_size_bytes, self, start_key, end_key)
+
+  def read(self, range_tracker):
+    filter_ = self.beam_options['filter_']
+    table = self._getTable()
+    read_rows = table.read_rows(start_key=range_tracker.start_position(),
+                                end_key=range_tracker.stop_position(),
+                                filter_=filter_)
+
+    for row in read_rows:
+      if range_tracker.stop_position() != b'':
+        if range_tracker.try_claim(row.row_key):
+          return
+      self.read_row.inc()
+      yield row
```

Comment:
This looks incorrect. You need to claim all records that you return here and you need to return when try_claim() returns false for the first time.

Also, are we missing some unit tests that should have captured this ?

Reply:
I checked other source, and, I think the problem is the 387 line:
```python
if range_tracker.stop_position() != b'':
```
so, I going to modify a little bit the code, because, f you get all the rows from bigtable the last bundle get `b''` as stop_position so, the try_claim get True and break the loop.

```python
def read(self, range_tracker):
    filter_ = self.beam_options['filter_']
    table = self._getTable()
    read_rows = table.read_rows(start_key=range_tracker.start_position(),
                                end_key=range_tracker.stop_position(),
                                filter_=filter_)

    for row in read_rows:
      try_claim = range_tracker.try_claim(row.row_key)
      # if you get all the rows from bigtable
      # the last bundle get `b''` as stop_position
      # so, the try_claim get True and break the loop.
      if range_tracker.stop_position() != b'':
        if try_claim:
          return
      self.read_row.inc()
      yield row
```

Reply:
It's the opposite. If try_claim() returns true the current reader owns that set of records and must return that. If try_claim() returns fails the current reader should return right away without returning any records that it failed to claim. The reader should not return any records that it didn't claim. See https://beam.apache.org/documentation/io/developing-io-python/

Reply:
So, I changed the code a little bit, but, have problems with the try_claim method. This uses the start_position and stop_position as a range to compare if the row is in between, the problem is that bigtable gets me a bundle with a empty stop_position. When try_claim checks the range of the last bundle the start_position is greater than the stop position and the try_claim returns false on all the rows within that range.

so, the try_claim did not work for the last bundle.

thats why i added this condition.
```python
      if range_tracker.stop_position() != b'':
```
example of the read:
```python
def read(self, range_tracker):
    filter_ = self.beam_options['filter_']
    table = self._getTable()
    read_rows = table.read_rows(start_key=range_tracker.start_position(),
                                end_key=range_tracker.stop_position(),
                                filter_=filter_)

    for row in read_rows:
      try_claim = range_tracker.try_claim(row.row_key)
      # if you get all the rows from bigtable
      # the last bundle get `b''` as stop_position
      # so, the try_claim get False and break the loop.
      if range_tracker.stop_position() != b'':
        if not try_claim:
          return
      self.read_row.inc()
      yield row
```

Reply:
I don't think we can skip try claim for any record (including the last bundle). Please note the not following this algorithm correctly can result in serious data correction (loss/duplication) issues. 

"When try_claim checks the range of the last bundle the start_position is greater than the stop position ": Why would that be the case ? Does "LexicographicKeyRangeTracker" somehow have to be modified to be usable for the BT source ?

Reply:
We need to figure out how to properly code for `LexicographicKeyRangeTracker(<range_start>, <infinity>)`.  The Java `ByteKeyRange` has some special handling for `empty` as the end key, and ideally `LexicographicKeyRangeTracker` should have similar behavior.